2020/02/21/consolidate-multiple-sftp-accounts-on-one-master-account.md

```diff
@@ -0,0 +1,154 @@
+chown -R root:root /home/$USERNAME
+```
+
+As the number of facilities increases, additional communication points are added. If the permissible amount of points is exceeded on the app by even one, it will not continue its service until more blocks are paid for. The implementation of this process will directly benefit our client.
+
+### Multiple Accounts Mounts to One Account
+
+The goal here is to point the data from many facilities to one location. Yet, using a single account and path for multiple sites’ data will result in a breach in security and privacy. However, mounting the receiving path of a facility’s data onto a single master account to then “mount point” with a unique facility name takes care of this issue. The process continues on to consolidate files from individual paths on a master account in one place where the application picks up messages for further processing.
+
+
+Both the SFTP accounts as well as master account should be attached to one group. This will permit individual accounts to write on the master account-mounted path. In turn, the master account can read files from the same location. This location now has administrative rights of both the SFTP user as well as shared group permission. Group permission of the mounted folder is set to sftpgroup and user permission is set to facility account.
+
+
+A script is written to automate the list of actions in order to create an SFTP account, mount it at the master account path, and add fstab entries to save the mount in the case of a reboot. The script not only saved much time, but also avoided human error when creating accounts for all facilities in addition to future usage.
+
+* Create master account, sftpgroup and mount paths
+
+```bash
+
+# create master user account
+$ adduser master
+$ passwd master
+
+# Add master user to sftpgroup group
+usermod -a -G sftpgroup master
+getent group | grep master
+
+# MOUNT_PATH and sub folders
+mkdir -p /home/master/MOUNT_PATH/{Input,Pickup,Backup,Archive}
+chown -R master:master /home/master/MOUNT_PATH
+```
+
+* Automated script to create sftp account and mount folders
+
+```bash
+#!/bin/bash
+
+# sh create_sftp_account.sh sftpfacilityone FACILITY_ONE
+
+USERNAME=$1
+FACILITY_NAME=$2      
+
+# Create SFTP account and add to sftpgroup 
+useradd $USERNAME
+usermod -g sftpusers -s /usr/sbin/nologin $USERNAME
+usermod -a -G sftpgroup $USERNAME
+mkdir -p /home/$USERNAME
+chown root:root /home/$USERNAME
+mkdir -p /home/$USERNAME/INPUT_PATH
+
+# Create path specific to facility in master account Input folder
+mkdir -p /home/master/MOUNT_PATH/Input/$FACILITY_NAME
+chown -R $USERNAME:sftpgroup /home/master/MOUNT_PATH/Input/$FACILITY_NAME
+
+
+# Mount sftp account into master account and set permissions
+mount --bind /home/master/MOUNT_PATH/Input/$FACILITY_NAME /home/$USERNAME/INPUT_PATH
+chmod g+s /home/$USERNAME/INPUT_PATH
+chown -R $USERNAME:sftpgroup /home/$USERNAME/INPUT_PATH
+setfacl -d -m g::rwx /home/$USERNAME/INPUT_PATH
```

Comment:
Is this doing anything different than the more common `chmod g=rwx`?

I'm wondering why it is separate from the earlier `chmod g+s` -- it seems they could be combined into `chmod g+rwsx`.

Is there something here I'm missing?

Reply:
@jonjensen 
The idea is to preserve the `sftpgroup` group permission for the files created in future. So setfacl used to set rwx permission to default group. But I have experimented the behavior of both ways and confirmed that we are good to use `chmod g+rwsx` since files created by root retained `sftpgroup`. 

```
* create folders to experiment

root@sl67:/home/sftptest# mkdir SETFACL
root@sl67:/home/sftptest# mkdir CHMOD
root@sl67:/home/sftptest# chown -R $sftpuser:sftpgroup /home/$sftpuser/{SETFACL,CHMOD}

* set permission by **setfacl** way 

root@sl67:/home/sftptest# chmod g+s /home/$sftpuser/SETFACL
root@sl67:/home/sftptest# setfacl -d -m g::rwx /home/$sftpuser/SETFACL
root@sl67:/home/sftptest# getfacl /home/$sftpuser/SETFACL
getfacl: Removing leading '/' from absolute path names
# file: home/sftptest/SETFACL
# owner: sftptest
# group: sftpgroup
# flags: -s-
user::rwx
group::r-x
other::r-x
default:user::rwx
default:group::rwx
default:other::r-x

* set permission by **chmod** way

root@sl67:/home/sftptest# chmod g=rwxs /home/$sftpuser/SETFACL
root@sl67:/home/sftptest# chmod g=rwxs /home/$sftpuser/CHMOD/
root@sl67:/home/sftptest# getfacl /home/$sftpuser/CHMOD
getfacl: Removing leading '/' from absolute path names
# file: home/sftptest/CHMOD
# owner: sftptest
# group: sftpgroup
# flags: -s-
user::rwx
group::rwx
other::r-x

* Experiment new file group permission 

root@sl67:/home/sftptest# touch SETFACL/setfacl.txt CHMOD/chmod.txt
root@sl67:/home/sftptest# ls -lrt SETFACL/ CHMOD/
SETFACL/:
total 0
-rw-rw-r-- 1 root sftpgroup 0 Feb 25 02:26 setfacl.txt

CHMOD/:
total 0
-rw-r--r-- 1 root sftpgroup 0 Feb 25 02:26 chmod.txt
```

Reply:
:+1: